README.md
# cs-327-rogue-game by David bone

## Assignment 1.10

Here is my final assignment for my Rouge-like game.

For this assignment, I decided to flesh out the 'value' aspect of objects by creating a buy and sell mechanic, complete with PC money and a town you can visit to spend said money on loot or potions to heal yourself.

To sell an item, from the main screen, press 's'. You are them prompted for a carry slot to sell from. This method works like the delete method, however it increments your money with the value of the sold object

To spend your money, from the main screen, press 'c'. This teleports you to a town with two places to visit, the Armor store and Thielen student health center. To enter a store, use the same movement keys from the game to move along the street. Then enter a door of one of the stores. Press Escape to leave town.

In the armor shop, you are presented with 3 randomly generated items that you can purchase for the displayed cost. Buying an item places it in the first available carry spot. Press escape to leave the store.

In Thielen, you can buy health items to immediately regenerate health. The minor health potion regenerates 50 health, the major health potion regenerates 100 health, and the coveted Clone Cone brings you to max health. Press escape to leave Thielen.

USAGE:

    >make
    > ./rlg327 --nummon <desired integer of monsters>

Tested on Ubuntu

PS: thank you for all your hard work as a TA for this class. It has been a very fun class and my favorite of the semester by far.

## Assignment 1.09

The penultimate assignment had me adding many object related commands to the dungeon. All now work as outlined in the assignment pdf. All must be used when looking at the dungeon and not having any other commands open. For instance, when the game is prompting the user for a carry slot after pressing the 'w' command, you cannot immediately look at the inventory, you must first press esc to cancel the wear command. All commands may be cancelled or exited using escape.

At the bottom of the screen, the player health will always be displayed, followed by the last damaged monster or the first monster in the dungeon array if no monster has been damaged. All damage, speed and health calculations work as outlined.

The game is won by killing a BOSS (spongebob squarepants), but that may be very difficult, as the BOSSes are rare. Once this happens, the game displays the 'crown' win graphic and then exits.

USAGE:

    >make
    > ./rlg327 --nummon <desired integer of monsters>

Tested on Ubuntu

## Assignment 1.08

This assignment had me using the new file parsing to add different types of monsters and objects to the dungeon.

I began by using Professor Sheaffer's parser for both objects and monsters. This took some time to read through and understand how to fit with my current architecture.

The dungeon works as it did in Assignment 1.06, but now it renders the new monsters and objects into the dungeon space.

Both the monsters and objects include the symbol, color, and abilities specified when generated. All other values are stored in their objects within the dungeon architecture. This includes rolling the values that are meant to be rolled at generation.

I also fixed an error I had in 1.06 where I had mapped the fog of war toggle to 't' instead of 'f'. Everything should work as expected.

I opted not to include any optional functionality in this assignment. Thank you!

USAGE:

    >make
    > ./rlg327 --nummon <desired integer of monsters>

Tested on Ubuntu

## Assignment 1.07

This assignment consisted of parsing a text document for information on various monsters. I did not implement the optional Item definitions.

When ran, the program searches for a file called "monster_desc.txt" in the home folder of this assignment, stores the data of all complete monsters in a vector, prints each to the terminal and exits.

The order in which the program prints to terminal is inline with that of the table in the assignment PDF. (I noticed that this order was different that the order the example output was in, notably, the symbol parameter was out of place. I formatted my output against the table.)

If a monster is missing a parameter or has more than one definition of a parameter, it is thrown out and the next monster is read.

USAGE:

    >make
    > ./rlg327

Tested on Ubuntu

## Assignment 1.06

There is now a default fog of war view in the game. This view is toggled by 't' during the players turn. You do not waste a turn to do this.

There is also now now a teleportation key 'g' that activates a selection window, with the new location denoted with '*'. You are allowed to teleport into rock, but not immutable rock. If a rock is immutable, you will not be able to select it.

Pressing 'r' while selecting a teleportation location will teleport the PC to a random non-immutable cell.

USAGE:

    >make
    > ./rlg327 --nummon <desired integer of monsters>

Tested on Ubuntu

## Assignment 1.05

I have now added user input into the dungeon game. I have added all Keys that were specified.
The terminal output is now using Ncurses rather than stdio. This took some reworking of my print dungeon method but in the end, it worked out really well

If a key is pressed that is not one specified, the turn is saved and the header prompts the user that the input was not recognized. This goes for the main screen, not the monster list screen.

I have used the header to tell the user what action has just taken place. The bottom two lines on the dungeon screen are place holders for future assignments.

NOTE: I have not updated the save or load features, per Professor Sheaffer's instruction

USAGE:

    >make
    > ./rlg327 --nummon <desired integer of monsters>

Tested on Ubuntu

Halfway there!! (with assignment 1 at least...)


## Assignment 1.04

This assignment added monsters and allowed those monsters to move. This utilized a priority queue using the heap.c file.
I also had a memory issue when using my pathfinding algorithm from last week, so I decided to scrap it and use Professor Sheaffer's logic.

I separated the monsters into a tunneling and non-tunneling method. If a monster was intelligent, it always used the pathfinding algorithm. If a monster was not intelligent, then it used a simple Pythagorean Theorem method to find its way to the PC

If a monster was Telepathic, it always used the current location of the PC to path-find. If it was not, I used a memory variable as well as a islineofsight method to determine if or where to move.

I decided to not implement logic to have the character move.

USAGE:

    >make
    > ./rlg327 --nummon <desired integer of monsters>

If --nummon is not used, the default value is 10.

Tested on Ubuntu

Thank you!

## Assignment 1.03

This assignment adds distance map generation for both tunneling and non-tunneling. Both are automatically output when the program is ran.

All argument logic still applies.

USAGE:

    >make
    > ./rlg327 <arguments>

NOTE: tested on Ubuntu

Thank you!

## Assignment 1.02

This assignment adds file read and write capabilities to the file DungeonGenerator.c.

It always reads or writes a file from the path

    $HOME/.rlg327/dungeon

I have included the following commandline arguments:

    "--load" => loads the dungeon at said path and displays it
    "--save" => saves dungeon to said path before exiting

If now "--load" argument is set, it will generate a new dungeon using the algorithm I created last week. You are able to use both arguments in one line in any order. This will rewrite the loaded file every time it is ran.

USAGE:

    >make
    > ./rlg327 <arguments>

NOTE: I tested this program on Ubuntu and it was fully functional. On the pyrite server, I {Segmentation fault(core dumped)} error. I assume this is something to do with creating a .rlg327 folder in home. It is fully functional in Ubuntu.

ALSO NOTE: The terminal window must be AT LEAST 82 x 24 as I have a border around the dungeon output. (maybe I should delete the border...)


Thank you!





## Assignment 1.01

This is the First assignment in the Rogue Game project! Booyah!

So far I have only created one file: DungeonGenerator.c. This file randomly generates a dungeon according to the guidelines listed in the assignment and prints it to the console. It was tested on the pyrite server(see CHANGELOG.md).

Some things to note:
   1. The program first randomly generates 6-10 rooms of varying size and places them inside the dungeon such that no rooms directly touch and there is a band of 1 cell thick around the perimeter of the dungeon data structure

   2. I then use a mathematical distance algorithm to connect all rooms with corridors (I can see that it is ugly, I will improve it later).

   3. I finally randomly place an UP staircase and a DOWN staircase in any of the rooms. I understand that it is not allowed to go into a corridor, however in professor Sheaffer's example, he has a staircase in corridors

I have implemented a simple hardness quality in all of the cells, but have not created a dynamically allocated room array (or perhaps I have, I am not exactly sure)

USAGE:

    >make
    >./generator.out

NOTE: The terminal window must be AT LEAST 82 x 24 as I have a border around the dungeon output.

Enjoy!


CHANGELOG.md
2/7/2021
    => Create the git repository for the project.
    => Began working on DungeonGenerator.c

2/8/2021
    => Rendered a dummy dungeon in the console.
    => Realized that I cant just use a 2d array of chars and I have to use an array of objects/cells. Looks like that will be a tomorrow project.
2/9/2021
    => Restructured the dungeon as a 2d array of cell structs
    => used an enum type to show cell type.
    => implemented random room array generation (and used a struct for each room)
    => inputted all rooms into the dungeon structure
    => created logic for checking if any of the rooms were out of bounds, touching each other, overlapping, etc.
    => fixed issue with rand() generating the same numbers.
    => discovered that when I use srand() with the current time, it generates the same dungeon within each second.
2/9/2021
    => began work on corridor generation
    => changed the room generation size, number etc.
    => used mathematical distance to pave paths from room 0 -> 1 -> ... -> maxrooms -> 0
    => fixed issue that caused sections of dungeon to be cut off
    => implemented staircase gen
    => tested on pyrite server and got errors pertaining to the math.h library not being recognized. (this is the same problem mentioned in Piazza @94). I updated my make file to include -lm and it worked perfectly on the pyrite server.
    => update README.md and CHANGELOG.md
    => created zip and turned in assignment1.01
2/16/2021
    => added define statements to easily change dungeon parameters.
    => created load.c which currently take a file at $HOME/.rlg327/dungeon and reads The first 3 parameters
2/17/2021
    => added corridor, room and player character load from file.
    => added full load from file
    => combined load.c and dungeongenerator.c into one file.
    => changed dungeon to be a struct that included room information
    => added write method to write to a file
    => made dungeon a global variable
    => added command line argument support
2/18/2021
    => tested on Ubuntu
    => update README.md and CHANGELOG.md
    => created zip and turned in assignment1.02
2/22/2021
    => copied professor Sheaffer's algorithm to base my pathfinder logic.
    => created new file pathfinding.c
2/23/2021
    => created .h file to store all definitions for all files
    => changed hardness and other vars from signed to unsigned
    => correctly added non-tunneling path generation
    => correctly added tunneling path generation
    => tested on Ubuntu
2/24/2021
    => update README.md and turned in assignment1.03
3/2/2021
    => Created a monster struct and used an array within dungeon to store them
    => created monster.c and init_monsters method.
    => wrote logic to randomly determine type of monster and print hex numbers
    => updated path finding methods to take in a starting position
    => began fixing monsters and PC being stored in the grid.
3/2/2021
    => Had an issue with memory allocation with my path finding. I scrapped mine and used Sheaffer's
    => Wrote logic to move non tunneling monsters.
3/3/2021
    => Wrote logic to move tunneling monsters.
    => Began work on the Heap that runs the game
    => Changed the PC character to be a monster type
    => fixed issue where nontunneling monsters could not kill pc
    => added game over text.
    => Updated CHANGELOG.md and README.md
    => Tested on Ubuntu
    => Turned in Assignment1.04
3/7/2021
    => Removed border around dungeon display (finally)
3/8/2021
    => added ncurses into my project
    => rewrote my printdungeon to output with ncurses
    => added logic to move the player with the given keys
3/9/2021
    => added support to go up and down staircases and to generate a new dungeon each time
    => added rest and quit features
    => added monster list
    => created paginator for the monster list and added keys to control it
3/22/2021
    => ported to C++
    => changed make file
    => changed structs to classes and separated npc and pc classes
    => implemented fog of war
    => implemented new fog print method and fog toggle
    => created teleportation method
    => added random toggle in teleportation method
    => Tested on Ubuntu
    => Turned in Assignment 1.06
3/30/2021
    => Temporarily disabled game main method
    => added file reading
    => added dictionary to test which keyword was being used
    => implemented logic to skip an incomplete or a poorly formatted monster
    => printed to terminal
    => stored all monsters in a vector
    => Tested on Ubuntu
    => Turned in Assignment 1.07
4/5/2021
    => added Professor Sheaffer's parser into mine
    => added dice and definitions into my project
    => updated my npc class to include the new variables included
    => created an object class to store objects
    => added an object array in the dungeon class
    => reworked init_monsters method to create an array of monsters using the monsters description array vector
    => implemented logic to connect my path finding ai to the new monster architecture
4/6/2021
    => Added object initiation to the dungeon
    => fixed print method to print both new monsters and object
    => added color rendering
    => fixed npc typing to include the 5 new attributes
    => fixed unique monsters and artifacts
    => tested on Ubuntu
    => turned in Assignment 1.08
4/11/2021
    => added inventory and carry space to pc
    => added is picked_up boolean to object
    => implemented auto pick up for objects on the map
    => finished the 'w' equip command.
4/12/2021
    => finished the 't' unequip command.
    => finished the 'd' drop command
    => finished the 'x' delete command
    => finished the 'i' list inventory command
    => finished the 'e' carrying command
    => finished the 'I' inspect command
    => finished the 'L' Look command
    => added the speed of inventory
    => disabled NPC vs NPC
    => added hp logic for equipment
    => allowed enemies to hit you

main.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <math.h>

#include "dungeondefinitions.h"

heap_t h;
dungeon_type dungeon;
int nummon;

int player_turn(dungeon_type *d);
void io_init_terminal(void);
void end_game();

int main(int argc, char const *argv[])
{
    io_init_terminal();

    int boolsave = 0;
    int boolload = 0;
    nummon = 10;
    for (int i = 1; i < argc; i++)
    {
        if (strcmp(argv[i], "--save") == 0)
        {
            boolsave = 1;
        }
        else if (strcmp(argv[i], "--load") == 0)
        {
            boolload = 1;
        }
        else if (strcmp(argv[i], "--nummon") == 0)
        {
            nummon = (int)strtol(argv[i + 1], NULL, 10);
            i++;
        }
    }

    if (boolload == 1)
    {
        load();
    }
    else
    {
        generate();
    }

    dungeon.PC.alive = 1;
    dungeon.PC.nextturn = 0;
    dungeon.PC.sequencenumber = 0;
    dungeon.PC.speed = 10;
    dungeon.PC.to_string = '@';
    dungeon.PC.isNPC = 0;

    dungeon_type *d;
    d = &dungeon;

    if (boolsave == 1)
    {
        writetodisk();
    }

    h = init_monsters(d, nummon);
    char *header = "Welcome to David Bone's cs327 project!";
    for (int i = 0; i < strlen(header); i++)
    {
        mvaddch(0, i, header[i]);
    }
    run_game(d);
}

void run_game(dungeon_type *d)
{

    static monster *c;
    while ((c = heap_remove_min(&h)))
    {

        if (c->isNPC == 0)
        {
            int out = 0;
            while (!out)
            {
                out = player_turn(d);
            }

            if (c->alive == 0)
            {
                end_game();
                break;
            }
        }
        else if (c->type[TUNNELING] == '1')
        {
            move_monster_tunneling(d, c);
        }
        else
        {
            movemonsternontunneling(d, c);
        }

        c->nextturn = c->nextturn + (1000 / c->speed);
        if (c->alive == 1)
        {
            c->hn = heap_insert(&h, c);
        }
    }
}

int player_turn(dungeon_type *d)
{

    int out = 1;
    printDungeon(d);

    char *placeholder = "placeholder";
    for (int i = 0; i < strlen(placeholder); i++)
    {
        mvaddch(22, i, placeholder[i]);
    }
    for (int i = 0; i < strlen(placeholder); i++)
    {
        mvaddch(23, i, placeholder[i]);
    }

    refresh();

    int ch = getch();
    char header[100];

    uint8_t yNext = d->PC.pos.y;
    uint8_t xNext = d->PC.pos.x;

    switch (ch)
    {
    case '7':
    case 'y':
        sprintf(header, "Moved PC one cell to the Upper Left.");
        yNext--;
        xNext--;
        break;
    case '8':
    case 'k':
        sprintf(header, "Moved PC one cell Up.");
        yNext--;
        break;
    case '9':
    case 'u':
        sprintf(header, "Moved PC one cell to the Upper Right.");
        yNext--;
        xNext++;
        break;
    case '6':
    case 'l':
        sprintf(header, "Moved PC one cell to the Right.");
        xNext++;
        break;
    case '3':
    case 'n':
        sprintf(header, "Moved PC one cell to the Lower Right.");
        yNext++;
        xNext++;
        break;
    case '2':
    case 'j':
        sprintf(header, "Moved PC one cell Down.");
        yNext++;
        break;
    case '1':
    case 'b':
        sprintf(header, "Moved PC one cell to the Lower Left.");
        yNext++;
        xNext--;
        break;
    case '4':
    case 'h':
        sprintf(header, "Moved PC one cell to the Left.");
        xNext--;
        break;
    case '>':
        if (d->map[d->PC.pos.y][d->PC.pos.x].type == DOWN)
        {
            heap_delete(&h);
            dungeon = generate();
            h = init_monsters(&dungeon, nummon);
            clear();
            sprintf(header, "Moved down the stairs.");
            for (int i = 0; i < strlen(header); i++)
            {
                mvaddch(0, i, header[i]);
            }
            run_game(&dungeon);
            return 1;
        }
        else
        {
            sprintf(header, "Must be on downward stairs.");
        }
        break;
    case '<':
        if (d->map[d->PC.pos.y][d->PC.pos.x].type == UP)
        {
            heap_delete(&h);
            generate();
            h = init_monsters(&dungeon, nummon);
            clear();
            sprintf(header, "Moved up the Stairs.");
            for (int i = 0; i < strlen(header); i++)
            {
                mvaddch(0, i, header[i]);
            }
            run_game(&dungeon);
            return 1;
        }
        else
        {
            sprintf(header, "Must be on upward stairs.");
        }
        break;
    case '5':
    case ' ':
    case '.':
        sprintf(header, "Waited a turn.");
        break;

    case 'm':
    {
        int numPages = nummon / 23 + 1;
        int curPage = 1;
        int done = 0;
        while (!done)
        {
            char monster_string[100];
            sprintf(monster_string, "List of Monsters");
            clear();
            for (int i = 0; i < strlen(monster_string); i++)
            {
                mvaddch(0, i, monster_string[i]);
            }
            for (int i = (curPage - 1) * 22; i < nummon && i < 22 * curPage; i++)
            {
                if (d->monsters[i].alive == 1)
                {
                    int Ydistance = d->PC.pos.y - d->monsters[i].pos.y;
                    int Xdistance = d->PC.pos.x - d->monsters[i].pos.x;
                    char NS[6];
                    char WE[5];
                    if (Ydistance < 0)
                    {
                        sprintf(NS, "south");
                        Ydistance = Ydistance * -1;
                    }
                    else
                    {
                        sprintf(NS, "north");
                    }

                    if (Xdistance < 0)
                    {
                        sprintf(WE, "east");
                        Xdistance = Xdistance * -1;
                    }
                    else
                    {
                        sprintf(WE, "west");
                    }
                    sprintf(monster_string, "%c, %d %s and %d %s", d->monsters[i].to_string, Ydistance, NS, Xdistance, WE);
                }
                else
                {
                    sprintf(monster_string, "%c, dead", d->monsters[i].to_string);
                }

                for (int j = 0; j < strlen(monster_string); j++)
                {
                    mvaddch((i % 22) + 1, j, monster_string[j]);
                }
            }

            sprintf(monster_string, "Page %d of %d", curPage, numPages);
            for (int i = 0; i < strlen(monster_string); i++)
            {
                mvaddch(23, i, monster_string[i]);
            }
            refresh();
            int ch2 = getch();
            switch (ch2)
            {
            case 27: //escape key
                done = 1;
                break;
            case 259: //up arrow
                if (curPage > 1)
                {
                    curPage--;
                }
                break;
            case 258: //down arrow
                if (curPage < numPages)
                {
                    curPage++;
                }
                break;
            }
        }

        out = 0;
        break;
    }

    case 'Q':
        clear();
        endwin();
        exit(0);
    default:
        sprintf(header, "Unknown Key Input: %d", ch);
        out = 0;
        break;
    }
    if (d->PC.pos.x != xNext || d->PC.pos.y != yNext)
    {
        if (d->map[yNext][xNext].type == ROCK)
        {
            sprintf(header, "Unable to move, wall is there");
        }
        else
        {
            d->PC.pos.x = xNext;
            d->PC.pos.y = yNext;
        }
    }

    clear();

    for (int i = 0; i < strlen(header); i++)
    {
        mvaddch(0, i, header[i]);
    }

    return out;
}

void io_init_terminal(void)
{
    initscr();
    raw();
    noecho();
    curs_set(0);
    keypad(stdscr, TRUE);
}

void end_game()
{
    char *closing = "Game OVER! You died. Press any button to continue";

    for (int i = 0; i < strlen(closing); i++)
    {
        mvaddch(0, i, closing[i]);
    }
    refresh();
    int ch = getch();
    endwin();
    exit(0);
}
